web_poet/requests.py

```diff
@@ -0,0 +1,132 @@
+    providing an optional ``request_downloader`` callable.
+
+    Providing the request implementation by dependency injection would be a good
+    alternative solution when you want to avoid setting up :mod:`contextvars`
+    like ``web_poet.request_backend_var``.
+
+    In any case, this doesn't contain any implementation about how to execute
+    any requests fed into it. When setting that up, make sure that the downloader
+    implementation returns a :class:`~.HttpResponse` instance.
+    """
+
+    def __init__(self, request_downloader: Callable = None):
+        self.request_downloader = request_downloader or _perform_request
+
+    async def request(
+        self,
+        url: str,
+        *,
+        method: str = "GET",
+        headers: Optional[Headers] = None,
+        body: Optional[Body] = None,
+    ) -> HttpResponse:
+        """This is a shortcut for creating a :class:`HttpRequest` instance and executing
+        that request.
+
+        A :class:`~.HttpResponse` instance should then be returned.
+
+        .. warning::
+            By convention, the request implementation supplied optionally to
+            :class:`~.HttpClient` should return a :class:`~.HttpResponse` instance.
+            However, the underlying implementation supplied might change that,
+            depending on how the framework using **web-poet** implements it.
+        """
+        headers = headers or {}
+        body = body or b""
+        req = HttpRequest(url=url, method=method, headers=headers, body=body)
+        return await self.request_downloader(req)
+
+    async def get(self, url: str, *, headers: Optional[Headers] = None) -> HttpResponse:
+        """Similar to :meth:`~.HttpClient.request` but peforming a ``GET``
+        request.
+        """
+        return await self.request(url=url, method="GET", headers=headers)
+
+    async def post(
+        self,
+        url: str,
+        *,
+        headers: Optional[Headers] = None,
+        body: Optional[Body] = None,
+    ) -> HttpResponse:
+        """Similar to :meth:`~.HttpClient.request` but performing a ``POST``
+        request.
+        """
+        return await self.request(url=url, method="POST", headers=headers, body=body)
+
+    async def batch_requests(
+        self, *requests: HttpRequest
+    ) -> List[Union[HttpResponse, Exception]]:
+        """Similar to :meth:`~.HttpClient.request` but accepts a collection of
+        :class:`~.HttpRequest` instances that would be batch executed.
+
+        If any of the :class:`~.HttpRequest` raises an exception upon execution,
+        they will be suppressed. Instead, the actual exception is returned
+        alongside any successful :class:`~.HttpResponse`. This enables salvaging
+        any usable responses despite any possible failures.
+        """
+
+        coroutines = [self.request_downloader(r) for r in requests]
+        responses = await asyncio.gather(*coroutines, return_exceptions=True)
```

Comment:
What do you think about exposing return_exceptions option? It could make sense to keep the interface similar to asyncio.gather. I also wonder if we should keep False as a default; it may be more intuitive. Otherwise, user would always need to add some code to the PO itself to handle the exceptions; with return_exceptions=False an error would be just raised & propagated, which may be a right behavior in many cases.

Reply:
Exposing `return_exceptions` would be a good idea, as it could prevent some additional effort from the the user to manually tweak the behavior. 👍 Updated this on https://github.com/scrapinghub/web-poet/pull/22/commits/bc4baea396de7227ff0dacb1db33c1f2b2b165a0

---

Regarding on having its _**default**_ value set to `False`: I do agree that it would be consistent with the underlying `asyncio.gather()` API default expectations. 

However, I'm not quite sure if we're encouraging the right behavior to the user. For our use case, executing HTTP requests incurs more cost than other coroutine operations that asyncio could have. As such, we'd be interested in encouraging the behavior of salvaging any usable HTTP Responses that the user can. Having the exceptions err up when `return_exceptions=False` would lose track of usable responses altogether.

Moreover, it doesn't promote future possibilities like retrying a specific failed request from a `batch_request()` since the developer doesn't know which request had the issue. 

Let me know what you think. :)

Reply:
Hi @kmike , as discussed offline last week, we'll be having the _**default**_ value of `return_exceptions` set to `False`.

This leads to a more meaningful exception message that explicitly logs `HttpRequestError` as opposed to something like _"css method not found in ..."_ since the developer has not automatically handled the returned exception. Raising them by default makes the developer more aware on how to handle them.

Updated in https://github.com/scrapinghub/web-poet/pull/22/commits/80f95a8f9a882e4fb3579f546e10bd548016e172.